### PR TITLE
Implement gossip v1.1

### PR DESCRIPTION
**Pull request by Nashatyrev**
## PR Description

Implementation of Gossip 1.1 protocol: 
https://github.com/libp2p/specs/blob/master/pubsub/gossipsub/gossipsub-v1.1.md

Limitation: 
- [Peer Exchange](https://github.com/libp2p/specs/blob/master/pubsub/gossipsub/gossipsub-v1.1.md#prune-backoff-and-peer-exchange) implemented without 'Signed peer records' (which is optional)

Public API changes: 
- Subscriber validation result can also be `Ignore` when message is valid but shouldn't be propagated
- Much more router and score parameters (totally >50 params) can now be passed to the `GossipRouter`
- Among these parameters there is `GossipParamsV1_1.connectCallback` which is invoked whenever `GossipRouter` wants the external peer manager to honor and connect to specific peers 
- Replace `ProtocolBinding` `announce+matcher` properties with new `ProtocolDescriptor` class (https://github.com/libp2p/jvm-libp2p/pull/108/commits/9b7a65c40fcdf8810a1409f8217c529a21a48661)
- Replace `ProtocolMatcher` (which is now a part of `ProtocolDescriptor`) with a simple predicate interface

## TODO: 
- [x] `1.0` and `1.1` compatibility
- [x] check Go Gossip 1.1 interoperability
- [x] Check and setup reasonable default score parameters 
- [x] Check Java interoperability

### CODE REVIEW

**Comment by cemozerr** [comment]
> * [Peer Exchange](https://github.com/libp2p/specs/blob/master/pubsub/gossipsub/gossipsub-v1.1.md#prune-backoff-and-peer-exchange) implemented without 'Signed peer records' (which is optional)

How do we implement Peer Exchange without Signed peer records? The following line from spec makes me think that PX is only possible when `PRUNE` control messages include the optional set of peers.

**Comment by Nashatyrev** [comment]
> 
> 
> > * [Peer Exchange](https://github.com/libp2p/specs/blob/master/pubsub/gossipsub/gossipsub-v1.1.md#prune-backoff-and-peer-exchange) implemented without 'Signed peer records' (which is optional)
> 
> How do we implement Peer Exchange without Signed peer records? The following line from spec makes me think that PX is only possible when `PRUNE` control messages include the optional set of peers.

`Signed Peer Records` seemed to be pretty large RFC so I would omit it for this release for the sake of resources economy. 

This version exchanges only `PeerId`s without address information which can also be helpful: if the discovery system knows the peer by id then the connection manager may favor those peers over others or add them additional reputation points or whatever (I'm not sure we have such mechanism at the moment)  

In short I would postpone SPR feature until it's actually wanted for Teku

File: src/main/kotlin/io/libp2p/pubsub/gossip/GossipRouter.kt [hunk 1]
```diff
@@ -98,104 +235,213 @@ open class GossipRouter : AbstractRouter() {
...
+    private fun emitGossip(topic: Topic, excludePeers: Collection<PeerHandler>) {
+        val ids = mCache.getMessageIds(topic)
+        if (ids.isEmpty()) return
+
+        val shuffledMessageIds = ids.shuffled(random).take(params.maxIHaveLength)
+        val peers = (getTopicPeers(topic) - excludePeers)
+            .filter { score.score(it) >= score.params.gossipThreshold && !isDirect(it) }
+
+        peers.shuffled(random)
+            .take(max((params.gossipFactor * peers.size).toInt(), coreParams.DLazy))
```
Comment (cemozerr) [inline src/main/kotlin/io/libp2p/pubsub/gossip/GossipRouter.kt:398]: I think this should be:
```suggestion
            .take(min((params.gossipFactor * peers.size).toInt(), coreParams.DLazy))
```
Since the spec says:
> From these peers, it randomly selects gossip factor peers with a minimum of D_lazy, and emits gossip to the selected peers.

Comment (cemozerr) [reply src/main/kotlin/io/libp2p/pubsub/gossip/GossipRouter.kt:398]: And it might be a good idea to add a test for it

File: src/main/kotlin/io/libp2p/pubsub/gossip/GossipRouter.kt [hunk 2]
```diff
@@ -98,104 +235,213 @@ open class GossipRouter : AbstractRouter() {
...
+                } else if (peers.size > coreParams.DHigh) {
+                    // too many mesh peers
+                    val sortedPeers = peers
+                        .shuffled(random)
+                        .sortedBy { score.score(it) }
+                        .reversed()
+
+                    val bestDPeers = sortedPeers.take(coreParams.DScore)
+                    val restPeers = sortedPeers.drop(coreParams.DScore).shuffled(random)
+                    val outboundCount = bestDPeers.count { it.isOutbound() }
```
Comment (cemozerr) [inline src/main/kotlin/io/libp2p/pubsub/gossip/GossipRouter.kt:333]: I think this should be:
```suggestion
                    val outboundCount = (bestDPeers + restPeers).count { it.isOutbound() }
```
As the spec says:
> **if the scoring plus random selection** does not result in enough outbound connections, then we replace the random and lower scoring peers in the selection with outboud connection peers.

File: src/main/kotlin/io/libp2p/pubsub/gossip/GossipRouter.kt [hunk 3]
```diff
@@ -98,104 +235,213 @@ open class GossipRouter : AbstractRouter() {
...
-                        prune(dropPeer, topic)
+
+                // drop underscored peers from mesh
+                peers.filter { score.score(it) < 0 }
+                    .forEach { prune(it, topic) }
+
+                if (peers.size < coreParams.DLow) {
+                    // need more mesh peers
+                    (getTopicPeers(topic) - peers)
+                        .filter { score.score(it) >= 0 && !isDirect(it) && !isBackOff(it, topic) }
```
Comment (cemozerr) [inline src/main/kotlin/io/libp2p/pubsub/gossip/GossipRouter.kt:320]: I know the spec does not say anything about this, but I'm curious if we should, on top of ignoring peers with a negative score, favor peers with a better score.

**Comment by cemozerr** [comment]
I could not find any modifications about Explicit Peering Agreements. I'm assuming it's not necessary for Eth2.0/Teku, so implementing it is not a priority for now?

File: src/main/kotlin/io/libp2p/pubsub/gossip/GossipRouter.kt [hunk 4]
```diff
@@ -98,104 +235,213 @@ open class GossipRouter : AbstractRouter() {
...
-                    (getTopicPeers(topic) - peers).shuffled(random).take(D - peers.size).forEach { newPeer ->
-                        peers += newPeer
-                        graft(newPeer, topic)
-                    }
-                } else if (peers.size > DHigh) {
-                    peers.shuffled(random).take(peers.size - D).forEach { dropPeer ->
-                        peers -= dropPeer
-                        prune(dropPeer, topic)
+
+                // drop underscored peers from mesh
```
Comment (cemozerr) [inline src/main/kotlin/io/libp2p/pubsub/gossip/GossipRouter.kt:313]: Should we not be doing the same thing to fanout topic and peers?

File: src/main/kotlin/io/libp2p/pubsub/gossip/GossipRouter.kt [hunk 5]
```diff
@@ -65,27 +96,133 @@ open class GossipRouter : AbstractRouter() {
...
+        val iWant = msg.messageIDsList - seenMessages.keys
+        val asked = iAsked.computeIfAbsent(peer) { AtomicInteger() }
+        val maxToAsk = max(0, min(iWant.size, params.maxIHaveLength - asked.get()))
+        asked.addAndGet(iWant.size)
+        iWant(peer, iWant.shuffled(random).subList(0, maxToAsk))
+    }
+
+    private fun handleIWant(msg: Rpc.ControlIWant, peer: PeerHandler) {
+        val peerScore = score.score(peer)
+        if (peerScore < score.params.gossipThreshold) return
```
Comment (cemozerr) [inline src/main/kotlin/io/libp2p/pubsub/gossip/GossipRouter.kt:213]: is IWant message considered a gossip message as well?

File: src/main/kotlin/io/libp2p/pubsub/gossip/GossipRouter.kt [hunk 6]
```diff
@@ -98,104 +235,213 @@ open class GossipRouter : AbstractRouter() {
...
                 }
-                submitGossip(topic, peers)
+
+                emitGossip(topic, peers)
             }
             fanout.entries.forEach { (topic, peers) ->
-                peers.removeIf { it in getTopicPeers(topic) }
-                val needMore = D - peers.size
+                peers.removeIf {
+                    it !in getTopicPeers(topic) || score.score(it) < scoreParams.publishThreshold
```
Comment (cemozerr) [inline src/main/kotlin/io/libp2p/pubsub/gossip/GossipRouter.kt:365]: The spec suggests that `publishThreshold` is only used when flood publishing. Is fanout a type of flood publishing? 
> publishThreshold: when a peer's score drops below this threshold, self published messages are not propagated towards this peer when (flood) publishing.

File: src/main/kotlin/io/libp2p/pubsub/gossip/GossipScore.kt [hunk 7]
```diff
@@ -0,0 +1,213 @@
...
+            this@GossipScore.params.decayToZero,
+            params.MeshMessageDeliveriesCap
+        )
+        var meshFailurePenalty: Double by cappedDouble(0.0, this@GossipScore.params.decayToZero)
+        var invalidMessages: Double by cappedDouble(0.0, this@GossipScore.params.decayToZero)
+
+        fun inMesh() = joinedMeshTime > 0
+
+        fun meshTimeNorm() = min(
+            (if (inMesh()) curTime() - joinedMeshTime else 0).toDouble() / params.TimeInMeshQuantum.toMillis(),
```
Comment (cemozerr) [inline src/main/kotlin/io/libp2p/pubsub/gossip/GossipScore.kt:44]: The spec suggests that the _"The time in mesh is calculated lazily during the decay update to avoid a large number of calls to `gettimeofday`_". 

I might not be correctly understanding the lazy keyword concept in Kotlin, but it seems to me that we are not lazily updating `meshTime` during the decay update, but calling `System.currentTimeMillis()` at every `score()` call.

File: src/main/kotlin/io/libp2p/pubsub/gossip/GossipScore.kt [hunk 8]
```diff
@@ -0,0 +1,213 @@
...
+class GossipScore(
+    val params: GossipScoreParams = GossipScoreParams(),
+    val executor: ScheduledExecutorService,
+    val curTime: () -> Long
+) {
+
+    inner class TopicScores(val topic: Topic) {
+        private val params = topicParams[topic]
+
+        var joinedMeshTime: Long = 0
```
Comment (cemozerr) [inline src/main/kotlin/io/libp2p/pubsub/gossip/GossipScore.kt:27]: nit:
```suggestion
        var joinedMeshTimeInMillis: Long = 0
```

File: src/main/kotlin/io/libp2p/pubsub/gossip/GossipScore.kt [hunk 9]
```diff
@@ -0,0 +1,213 @@
...
+import java.util.concurrent.ScheduledFuture
+import java.util.concurrent.TimeUnit
+import kotlin.math.max
+import kotlin.math.min
+import kotlin.math.pow
+
+class GossipScore(
+    val params: GossipScoreParams = GossipScoreParams(),
+    val executor: ScheduledExecutorService,
+    val curTime: () -> Long
```
Comment (cemozerr) [inline src/main/kotlin/io/libp2p/pubsub/gossip/GossipScore.kt:21]: nit:
```suggestion
    val curTimeInMillis: () -> Long
```

File: src/main/kotlin/io/libp2p/pubsub/gossip/GossipScore.kt [hunk 10]
```diff
@@ -0,0 +1,213 @@
...
+
+        fun inMesh() = joinedMeshTime > 0
+
+        fun meshTimeNorm() = min(
+            (if (inMesh()) curTime() - joinedMeshTime else 0).toDouble() / params.TimeInMeshQuantum.toMillis(),
+            params.TimeInMeshCap
+        )
+
+        fun isMeshMessageDeliveriesActive() =
+            inMesh() && ((curTime() - joinedMeshTime).millis > params.MeshMessageDeliveriesActivation)
```
Comment (cemozerr) [inline src/main/kotlin/io/libp2p/pubsub/gossip/GossipScore.kt:49]: Spec again suggests this `curTime()` to be calculated lazily during the decay update.

File: src/main/kotlin/io/libp2p/pubsub/gossip/GossipRouter.kt [hunk 11]
```diff
@@ -65,27 +96,133 @@ open class GossipRouter : AbstractRouter() {
...
+        }
+
+        val iWant = msg.messageIDsList - seenMessages.keys
+        val asked = iAsked.computeIfAbsent(peer) { AtomicInteger() }
+        val maxToAsk = max(0, min(iWant.size, params.maxIHaveLength - asked.get()))
+        asked.addAndGet(iWant.size)
+        iWant(peer, iWant.shuffled(random).subList(0, maxToAsk))
+    }
+
+    private fun handleIWant(msg: Rpc.ControlIWant, peer: PeerHandler) {
```
Comment (cemozerr) [inline src/main/kotlin/io/libp2p/pubsub/gossip/GossipRouter.kt:211]: I believe we have to implement this:
> IWANT message responses are limited in the number of retransmissions to a certain peer; in gossipsub v1.0 the router always responds to IWANT messages when the message in the cache. In gossipsub v1.1 the router responds a limited number of times to each peer so that IWANT spam does not cause a signficant drain of resources.

Unless you implemented this somewhere else and I'm missing it.

File: src/main/kotlin/io/libp2p/pubsub/gossip/GossipRouter.kt [hunk 12]
```diff
@@ -65,27 +96,133 @@ open class GossipRouter : AbstractRouter() {
...
+        val peerScore = score.score(peer)
+        // we ignore IHAVE gossip from any peer whose score is below the gossip threshold
+        if (peerScore < score.params.gossipThreshold) return
+        if (peerIHave.computeIfAbsent(peer) { AtomicInteger() }.incrementAndGet() > params.maxIHaveMessages) {
+            return
+        }
+
+        val iWant = msg.messageIDsList - seenMessages.keys
+        val asked = iAsked.computeIfAbsent(peer) { AtomicInteger() }
+        val maxToAsk = max(0, min(iWant.size, params.maxIHaveLength - asked.get()))
```
Comment (cemozerr) [inline src/main/kotlin/io/libp2p/pubsub/gossip/GossipRouter.kt:206]: Is the number of maximum ids in an `IWANT` equal to the number of maximum ids in an `IHAVE` message?

File: src/main/kotlin/io/libp2p/pubsub/gossip/GossipRouter.kt [hunk 13]
```diff
@@ -98,104 +235,213 @@ open class GossipRouter : AbstractRouter() {
...
+    private fun prune(peer: PeerHandler, topic: Topic) {
+        setBackOff(peer, topic)
+        enqueuePrune(peer, topic)
+        if (mesh[topic]?.remove(peer) == true) {
+            notifyPruned(peer, topic)
+        }
+    }
+
+    private fun iWant(peer: PeerHandler, messageIds: List<MessageId>) {
+        messageIds.forEach { iWantRequests[peer to it] = curTime() }
```
Comment (cemozerr) [inline src/main/kotlin/io/libp2p/pubsub/gossip/GossipRouter.kt:417]: It seems like the spec is suggesting to track only one random message-id, not all of the message-ids.
> For each IHAVE advertisement which elicits an IWANT request, the router tracks a random message ID within the advertised set.

File: src/main/kotlin/io/libp2p/core/pubsub/PubsubApi.kt [hunk 14]
```diff
@@ -12,11 +12,33 @@ import kotlin.random.Random.Default.nextLong
...
+ */
+enum class ValidationResult {
+
+    /**
+     * The message is valid and should be propagated further
+     */
+    Valid,
+
+    /**
+     * The message is invalid
```
Comment (cemozerr) [inline src/main/kotlin/io/libp2p/core/pubsub/PubsubApi.kt:26]: nit: 
```suggestion
     * The message is invalid, should be rejected, and trigger penalty.
```

File: src/main/kotlin/io/libp2p/core/pubsub/PubsubApi.kt [hunk 15]
```diff
@@ -12,11 +12,33 @@ import kotlin.random.Random.Default.nextLong
...
+     */
+    Invalid,
+
+    /**
+     * The message is valid but should NOT be propagated further
```
Comment (cemozerr) [inline src/main/kotlin/io/libp2p/core/pubsub/PubsubApi.kt:31]: nit:
```suggestion
     * The message should not be propagated further but does not trigger a penalty.
```

File: src/main/kotlin/io/libp2p/pubsub/gossip/GossipRouter.kt (hunk 1 above)
Comment (Nashatyrev) [reply src/main/kotlin/io/libp2p/pubsub/gossip/GossipRouter.kt:398]: From my understanding of the spec `D_lazy` is the lower bound so if the `params.gossipFactor * peers.size` value is lower than `D_lazy` then the `D_lazy` value should be used. This is `max`. 
May be I missed something?

File: src/main/kotlin/io/libp2p/pubsub/gossip/GossipRouter.kt (hunk 2 above)
Comment (Nashatyrev) [inline src/main/kotlin/io/libp2p/pubsub/gossip/GossipRouter.kt:333]: @cemozerr seems like you deleted your comment here, but it was near correct
It should be
```kt
val outboundCount = (bestDPeers + restPeers)
    .take(coreParams.D)
    .count { it.isOutbound() }
```

File: src/main/kotlin/io/libp2p/pubsub/gossip/GossipRouter.kt (hunk 3 above)
Comment (Nashatyrev) [reply src/main/kotlin/io/libp2p/pubsub/gossip/GossipRouter.kt:320]: Do you mean sort the peers by score before taking desired number of peers? 
There is no such requirement and the reference Go implementation doesn't do it: https://github.com/libp2p/go-libp2p-pubsub/blob/master/gossipsub.go#L1283

I believe there could be some sense behind this: we don't want total score biasing, there should be some randomness. But I hope @vyzo might want to comment on this

**Comment by Nashatyrev** [comment]
> 
> 
> I could not find any modifications about Explicit Peering Agreements. I'm assuming it's not necessary for Eth2.0/Teku, so implementing it is not a priority for now?

We are handling explicit (or direct) peers via this Gossip parameter callback: https://github.com/libp2p/jvm-libp2p/blob/7fc89c886a04fddafc123c3dc14d19592cbd92ba/src/main/kotlin/io/libp2p/pubsub/gossip/GossipParams.kt#L80

Yep, the Gossip spec says that the gossip _router_ should establish check and manage connection to explicit peers. However I'm not sure if it makes sense to empower the pubsub router with such connection manager duties. I believe if/when the client application would like to manage such explicit peers

File: src/main/kotlin/io/libp2p/pubsub/gossip/GossipRouter.kt (hunk 4 above)
Comment (Nashatyrev) [reply src/main/kotlin/io/libp2p/pubsub/gossip/GossipRouter.kt:313]: Fanout peers are not in the mesh and we can't prune them, however we are also cleaning our fanout list from underscored peers: 
https://github.com/libp2p/jvm-libp2p/blob/7fc89c886a04fddafc123c3dc14d19592cbd92ba/src/main/kotlin/io/libp2p/pubsub/gossip/GossipRouter.kt#L365

File: src/main/kotlin/io/libp2p/pubsub/gossip/GossipRouter.kt (hunk 5 above)
Comment (Nashatyrev) [reply src/main/kotlin/io/libp2p/pubsub/gossip/GossipRouter.kt:213]: Yep, `IWANT` is a purely Gossip message

File: src/main/kotlin/io/libp2p/pubsub/gossip/GossipRouter.kt (hunk 6 above)
Comment (Nashatyrev) [reply src/main/kotlin/io/libp2p/pubsub/gossip/GossipRouter.kt:365]: Hmm, good question! 
That condition was taken from the Go implementation: https://github.com/libp2p/go-libp2p-pubsub/blob/master/gossipsub.go#L1440-L1442
But I also didn't find any mentions in the spec. 
@vyzo may be you can comment?

File: src/main/kotlin/io/libp2p/pubsub/gossip/GossipScore.kt (hunk 7 above)
Comment (Nashatyrev) [reply src/main/kotlin/io/libp2p/pubsub/gossip/GossipScore.kt:44]: The current implementation is very far from the optimal with regards to performance. `System.currentTimeMillis()` is absolutely not the bottleneck at the moment. 
I'm not sure why that is mentioned in the spec. 
May be @vyzo could comment on this?

File: src/main/kotlin/io/libp2p/pubsub/gossip/GossipRouter.kt (hunk 11 above)
Comment (Nashatyrev) [reply src/main/kotlin/io/libp2p/pubsub/gossip/GossipRouter.kt:211]: Good catch! 👍 
Missed that restriction

File: src/main/kotlin/io/libp2p/pubsub/gossip/GossipScore.kt (hunk 8 above)
Comment (Nashatyrev) [reply src/main/kotlin/io/libp2p/pubsub/gossip/GossipScore.kt:27]: If you don't mind I would change to `joinedMeshTimeMillis`

File: src/main/kotlin/io/libp2p/pubsub/gossip/GossipScore.kt (hunk 9 above)
Comment (Nashatyrev) [reply src/main/kotlin/io/libp2p/pubsub/gossip/GossipScore.kt:21]: If you don't mind I would change to `curTimeMillis`

File: src/main/kotlin/io/libp2p/pubsub/gossip/GossipScore.kt (hunk 10 above)
Comment (Nashatyrev) [reply src/main/kotlin/io/libp2p/pubsub/gossip/GossipScore.kt:49]: Commented above

File: src/main/kotlin/io/libp2p/pubsub/gossip/GossipRouter.kt (hunk 13 above)
Comment (Nashatyrev) [reply src/main/kotlin/io/libp2p/pubsub/gossip/GossipRouter.kt:417]: Right, that makes sense! Else the penalty can be too large for a single missed `IWANT`

File: src/main/kotlin/io/libp2p/pubsub/gossip/GossipRouter.kt (hunk 1 above)
Comment (cemozerr) [reply src/main/kotlin/io/libp2p/pubsub/gossip/GossipRouter.kt:398]: Yes, you're right. Sorry for the confusion.

File: src/main/kotlin/io/libp2p/pubsub/gossip/GossipRouter.kt (hunk 3 above)
Comment (vyzo) [reply src/main/kotlin/io/libp2p/pubsub/gossip/GossipRouter.kt:320]: Yes indeed, we want random selection!

File: src/main/kotlin/io/libp2p/pubsub/gossip/GossipRouter.kt (hunk 6 above)
Comment (vyzo) [reply src/main/kotlin/io/libp2p/pubsub/gossip/GossipRouter.kt:365]: Yeah, the publish threshold applies everywhere there is an actual publication.

File: src/main/kotlin/io/libp2p/pubsub/gossip/GossipScore.kt (hunk 7 above)
Comment (vyzo) [reply src/main/kotlin/io/libp2p/pubsub/gossip/GossipScore.kt]: It's very fast in most OSes, due to VDSO.

Comment (vyzo) [reply src/main/kotlin/io/libp2p/pubsub/gossip/GossipScore.kt]: But then again some OSes have a very slow implementation (a syscall) -- see OpenBSD.

Comment (vyzo) [reply src/main/kotlin/io/libp2p/pubsub/gossip/GossipScore.kt]: It is mentioned in the spec so that there is leeway to implement it this way, if you care to avoid calling `gettimeofday` in every score invocation -- it could be called potentially for every peer within the heartbeat, hence the reason for laziness.

Comment (vyzo) [reply src/main/kotlin/io/libp2p/pubsub/gossip/GossipScore.kt]: Another point that is relevant, with the lazy implementation we can avoid biasing the time in mesh by order of score invocation. This _could_ be an issue if you have a slow `gettimeofday` in your system.

**Comment by vyzo** [comment]
> Yep, the Gossip spec says that the gossip router should establish check and manage connection to explicit peers. However I'm not sure if it makes sense to empower the pubsub router with such connection manager duties. I believe if/when the client application needs such kind of dedicated peers they should be managed on a higher level and the Gossip router should be just notified that those peers are explicit. Again calling for @vyzo to comment on this

This is fine, you can implement it this way if you so desire. However, it should be pointed out that there should be some logic _somewhere_ for maintaining connections for explicit peering agreements. We implemented this way in the reference implementation because it was easy and it was requested as an explicit feature.

File: src/main/kotlin/io/libp2p/pubsub/gossip/GossipRouter.kt (hunk 11 above)
Comment (Nashatyrev) [reply src/main/kotlin/io/libp2p/pubsub/gossip/GossipRouter.kt:211]: Done: https://github.com/libp2p/jvm-libp2p/commit/94a02d0c9dae34d9fdc18ab79a61d99fc9560e49

File: src/main/kotlin/io/libp2p/pubsub/gossip/GossipRouter.kt (hunk 12 above)
Comment (Nashatyrev) [reply src/main/kotlin/io/libp2p/pubsub/gossip/GossipRouter.kt:206]: Yep, it's not the part of the spec, but here is the Go implementation doc on this param: 
https://github.com/libp2p/go-libp2p-pubsub/blob/master/gossipsub.go#L150-L155

Comment (Nashatyrev) [reply src/main/kotlin/io/libp2p/pubsub/gossip/GossipRouter.kt:206]: However I'm not sure about the logic of how this parameter is processed in Go. Seems slightly odd. Will revisit it again

**Comment by Nashatyrev** [comment]
There are still no reasonable average defaults for scoring params, and Go implementation left this onto client code. I.e. the default params mean _no scoring_

File: src/main/kotlin/io/libp2p/pubsub/gossip/GossipRouter.kt (hunk 12 above)
Comment (Nashatyrev) [reply src/main/kotlin/io/libp2p/pubsub/gossip/GossipRouter.kt]: Hm, now looks ok to me. Add a test for this and fix a bug

File: src/main/kotlin/io/libp2p/pubsub/gossip/GossipParams.kt [hunk 16]
```diff
@@ -10,62 +10,222 @@ import kotlin.math.min
...
+/**
+ * Parameters of Gossip 1.1 router
+ */
+data class GossipParams(
+    /**
+     * [D] sets the optimal degree for a GossipSub topic mesh. For example, if [D == 6],
+     * each peer will want to have about six peers in their mesh for each topic they're subscribed to.
+     * [D] should be set somewhere between [DLow] and [DHi].
+     */
     val D: Int = 3,
```
Comment (cemozerr) [inline src/main/kotlin/io/libp2p/pubsub/gossip/GossipParams.kt:32]: nit: Isn't this a bit low?

Comment (Nashatyrev) [reply src/main/kotlin/io/libp2p/pubsub/gossip/GossipParams.kt:32]: Right 👍 
`3` is some ancient default. `6` is in the Go impl
